lib/rgeo/impl_helper/valid_op.rb

```diff
@@ -0,0 +1,357 @@
+          return check unless check.nil?
+        end
+
+        # check closed
+        return TopologyErrors::UNCLOSED_RING unless is_closed?
+
+        # check more than 3 points
+        return TopologyErrors::TOO_FEW_POINTS unless num_points > 3
+
+        # check no self-intersections
+        check = check_no_self_intersections(self)
+        return check unless check.nil?
+
+        nil
+      end
+
+      def check_valid_polygon
+        # check coordinates are all valid
+        exterior_ring.points.each do |pt|
+          check = check_invalid_coordinate(pt)
+          return check unless check.nil?
+        end
+        interior_rings.each do |ring|
+          ring.points.each do |pt|
+            check = check_invalid_coordinate(pt)
+            return check unless check.nil?
+          end
+        end
+
+        # check closed
+        return TopologyErrors::UNCLOSED_RING unless exterior_ring.is_closed?
+        return TopologyErrors::UNCLOSED_RING unless interior_rings.all?(&:is_closed?)
+
+        # check more than 3 points in each ring
+        return TopologyErrors::TOO_FEW_POINTS unless exterior_ring.num_points > 3
+        return TopologyErrors::TOO_FEW_POINTS unless interior_rings.all? { |r| r.num_points > 3 }
+
+        # can skip this check if there's no holes
+        if interior_rings.size.positive?
+          check = check_consistent_area(self)
+          return check unless check.nil?
+        end
+
+        # check that there are no self-intersections
+        check = check_no_self_intersecting_rings(self)
+        return check unless check.nil?
+
+        # can skip these checks if there's no holes
+        if interior_rings.size.positive?
+          check = check_holes_in_shell(self)
+          return check unless check.nil?
+
+          check = check_holes_not_nested(self)
+          return check unless check.nil?
+
+          check = check_connected_interiors(self)
+          return check unless check.nil?
+        end
+
+        nil
+      end
+
+      def check_valid_multi_point
+        geometries.each do |pt|
+          check = check_invalid_coordinate(pt)
+          return check unless check.nil?
+        end
+        nil
+      end
+
+      def check_valid_multi_polygon
+        geometries.each do |poly|
+          return poly.invalid_reason unless poly.invalid_reason.nil?
+        end
+
+        check = check_consistent_area_mp(self)
+        return check unless check.nil?
+
+        # check no shells are nested
+        check = check_shells_not_nested(self)
+        return check unless check.nil?
+
+        nil
+      end
+
+      def check_valid_geometry_collection
+        geometries.each do |geom|
+          return geom.invalid_reason unless geom.invalid_reason.nil?
+        end
+
+        nil
+      end
+
+      ##
+      # Helper functions for specific validity checks
+      ##
+
+      def check_invalid_coordinate(pt)
+        return unless pt.x.nan? || pt.x.infinite? || pt.y.nan? || pt.y.infinite?
```

Comment:
```suggestion
        return if pt.x.finite? && pt.y.finite?
```

Should we also check for rational (not sure if it is an issue) and complex (`Numeric#real?`) as well ?

Reply:
I added the check for `real?` as well.